# Nextflow joint variant calling QC Pipeline with Automated Testing Integration

![Nextflow](https://img.shields.io/badge/nextflow-%E2%89%A522.10.1-brightgreen.svg)
![Docker](https://img.shields.io/badge/docker-ready-blue.svg)
![Testing](https://img.shields.io/badge/tests-passing-success)

## ðŸ“Œ Overview
This repository contains a Nextflow DSL2 implementation of a Joint Variant Calling pipeline. It is designed to demonstrate scalable, reproducible genomic analysis workflows using modern software engineering practices.

The pipeline follows the logic of the GATK variant discovery:

    Per-Sample Calling: Generating GVCFs from aligned BAM files.

    Consolidation: Merging GVCFs across the cohort.

    Joint Genotyping: Calling variants simultaneously across all samples to improve sensitivity for rare variants.

    Note: This is a practice repository created to demonstrate proficiency in Nextflow orchestration, module reusability, and containerized execution.

Pipeline Steps

The workflow consists of the following modular processes:

    1. Index BAMs: (samtools index) - Prepares input alignments.
    
    2. HaplotypeCaller: (GATK4) - Runs in GVCF mode on each sample independently (Scatter).

    3. CombineGVCFs: (GATK4) - Merges individual GVCFs into a single cohort GVCF (Gather).

    4. GenotypeGVCFs: (GATK4) - Performs the final joint genotyping step.
Repo Structure:

â”œâ”€â”€ data
â”‚Â Â  â”œâ”€â”€ bam
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ reads_father.bam
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ reads_mother.bam
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ reads_mother.bam.bai
â”‚Â Â  â”‚Â Â  â””â”€â”€ reads_son.bam
â”‚Â Â  â”œâ”€â”€ ref
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ intervals.bed
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ ref.dict
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ ref.fasta
â”‚Â Â  â”‚Â Â  â””â”€â”€ ref.fasta.fai
â”‚Â Â  â”œâ”€â”€ sample_bams.txt
â”‚Â Â  â””â”€â”€ samplesheet.csv
â”œâ”€â”€ genomics-1.nf
â”œâ”€â”€ genomics-2.nf
â”œâ”€â”€ genomics-3.nf
â”œâ”€â”€ joint-variant-calling-pipeline.nf
â”œâ”€â”€ modules
â”‚Â Â  â”œâ”€â”€ gatk
â”‚Â Â  â”‚Â Â  â”œâ”€â”€ haplotypecaller
â”‚Â Â  â”‚Â Â  â”‚Â Â  â”œâ”€â”€ main.nf
â”‚Â Â  â”‚Â Â  â”‚Â Â  â””â”€â”€ tests
â”‚Â Â  â”‚Â Â  â”‚Â Â      â”œâ”€â”€ main.nf.test
â”‚Â Â  â”‚Â Â  â”‚Â Â      â””â”€â”€ main.nf.test.snap
â”‚Â Â  â”‚Â Â  â””â”€â”€ jointgenotyping
â”‚Â Â  â”‚Â Â      â”œâ”€â”€ main.nf
â”‚Â Â  â”‚Â Â      â””â”€â”€ tests
â”‚Â Â  â”‚Â Â          â”œâ”€â”€ inputs
â”‚Â Â  â”‚Â Â          â”‚Â Â  â”œâ”€â”€ reads_father.bam.g.vcf
â”‚Â Â  â”‚Â Â          â”‚Â Â  â”œâ”€â”€ reads_father.bam.g.vcf.idx
â”‚Â Â  â”‚Â Â          â”‚Â Â  â”œâ”€â”€ reads_mother.bam.g.vcf
â”‚Â Â  â”‚Â Â          â”‚Â Â  â”œâ”€â”€ reads_mother.bam.g.vcf.idx
â”‚Â Â  â”‚Â Â          â”‚Â Â  â”œâ”€â”€ reads_son.bam.g.vcf
â”‚Â Â  â”‚Â Â          â”‚Â Â  â””â”€â”€ reads_son.bam.g.vcf.idx
â”‚Â Â  â”‚Â Â          â””â”€â”€ main.nf.test
â”‚Â Â  â””â”€â”€ samtools
â”‚Â Â      â””â”€â”€ index
â”‚Â Â          â”œâ”€â”€ main.nf
â”‚Â Â          â””â”€â”€ tests
â”‚Â Â              â”œâ”€â”€ main.nf.test
â”‚Â Â              â””â”€â”€ main.nf.test.snap
â”œâ”€â”€ nextflow.config
â”œâ”€â”€ nf-test.config


ðŸ§ª Testing & Validation

To ensure pipeline stability, this repository utilizes a basic testing strategy:

    Unit Tests: Each module (e.g., HaplotypeCaller) can be tested independently.

    Integration Test: The whole pipeline also can be tested in a single go.
